### PR TITLE
[3627] Reactify Approve/RequestPublish dialog

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/3627

### CODE REVIEW

File: static-assets/components/cstudio-common/common-api.js [hunk 1]
```diff
@@ -160,17 +160,17 @@ var nodeOpen = false,
          */
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js]: **Specifically** for common API: Revert the changes, add back what ever changes for this PR, disable the "reformat code" option prior to committing and then commit. Basically make sure only the changes of this PR come up here without all the code reformat (for this file, we need to do that on a stand-alone PR some other time).

File: ui/app/src/components/DateTimePicker.tsx [hunk 2]
```diff
@@ -23,15 +23,11 @@ import moment from 'moment-timezone';
 import AccessTimeIcon from '@material-ui/icons/AccessTime';
```
Comment (rart) [inline ui/app/src/components/DateTimePicker.tsx]: We need to consider the changes made to this API from the Audiences Panel PR.

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 3]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
+    items,
+    scheduling = 'now',
+    onClose
+  } = props;
 
   const [open, setOpen] = React.useState(true);
-  const [dialog, setDialog] = useReducer((a, b) => ({ ...a, ...b }), dialogInitialState);
+  const [dialog, setDialog] = useReducer((a, b) => ({ ...a, ...b }), {
+    ...dialogInitialState,
+    'scheduling': scheduling
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: `'scheduling': scheduling` => `scheduling`

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 4]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
+      ({ response }) => {
+        setPublishingChannels(response.availablePublishChannels);
+        setPublishingChannelsStatus('Success');
+        setSubmitDisabled(false);
+      },
+      ({ response }) => {
+        setPublishingChannelsStatus('Error');
+        setSubmitDisabled(true);
+      }
+    );
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Move into the useEffect(`=> HERE <=`, [...])

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 5]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
+    function setRef() {
+      const result = (
+        Object.entries({ ...checkedItems, ...checkedSoftDep })
+        .filter(([key, value]) => value === true)
+        .map(([key]) => key)
+      );
+      setSelectedItems(result);
+    }
+
+    setRef();
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Looks like you could simply run the function inside the effect without the additional function/invocation

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx (hunk 4 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishDialog.tsx]: @rart the detail in here is that `getPublishingChannels` is passed down to PublishForm, if getPublishingChannels fail it shows a `retry` button, that calls the function again. I can change it to something like `onGetPublishingChannelsFail` and call the function in the PublishDialog component, but I'd still need to have the function out from the useEffect. Please let me know what you think

File: static-assets/components/cstudio-common/common-api.js (hunk 1 above)
Comment (rart) [reply static-assets/components/cstudio-common/common-api.js]: Still not done.

File: ui/app/src/modules/Content/Publish/PublishForm.tsx [hunk 6]
```diff
@@ -176,28 +179,31 @@ function PublishForm(props: PublishFormProps) {
...
-      setInputs({ ...inputs, [name]: e.target.value })
-    }
+    } else if (e.target.type === 'textarea') {
+      setInputs({ ...inputs, [name]: e.target.value });
+    } else if (e.target.type === 'radio') {
+      const inputValue = e.target.value;
+      setInputs({ ...inputs, [name]: inputValue });
 
+      if (inputValue === 'now') {
+        setTimeout(() => {
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishForm.tsx]: What's the timeout for?

Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishForm.tsx]: There was a comment to reset the form values when changing from `Later` to `Now` (after 2 seconds)

File: static-assets/components/cstudio-common/common-api.js (hunk 1 above)
Comment (jvega190) [reply static-assets/components/cstudio-common/common-api.js]: done, please verify

File: ui/app/src/components/DateTimePicker.tsx [hunk 7]
```diff
@@ -95,80 +128,124 @@ interface timezoneType {
...
     switch (name) {
       case 'scheduledDate':
-        updatedDateTime.date(date.getDate());
-        updatedDateTime.month(date.getMonth());
-        updatedDateTime.year(date.getFullYear());
+        if (updatedDateTime._isValid) {
+          updatedDateTime.hours(timeMoment.toDate().getHours());
+          updatedDateTime.minutes(timeMoment.toDate().getMinutes());
+          updatedDateTime.seconds(timeMoment.toDate().getSeconds());
+          updatedDateTime.milliseconds(timeMoment.toDate().getSeconds());
```
Comment (rart) [inline ui/app/src/components/DateTimePicker.tsx]: Copy paste issue (getSeconds to set milliseconds).

Extract the .toDate to a const instead of repeatedly retrieving it. 
Place outside of the switch. You use it in both paths anyway.

File: ui/app/src/components/DateTimePicker.tsx [hunk 8]
```diff
@@ -95,80 +128,124 @@ interface timezoneType {
...
-    onChange?.(updatedDateTime);
-
-    onChangeDate?.(updatedDateTime.format(dateFormat));
-    onChangeTime?.(updatedDateTime.format(timeFormat));
+    if (updatedDateTime._isValid) {
+      onChange?.(updatedDateTime);
+    }
   };
 
   const handleTimezoneChange = () => (event: React.ChangeEvent<{}>, timezoneObj: any) => {
```
Comment (rart) [inline ui/app/src/components/DateTimePicker.tsx]: No point this being a creator instead of a straight function. 

Below, where this is used, the second param would only ever comes if the autocomplete sends these two a params.

File: ui/app/src/components/DateTimePicker.tsx [hunk 9]
```diff
@@ -204,8 +288,8 @@ const DateTimePicker = withStyles(dateTimePickerStyles)((props: DateTimePickerPr
         controls.includes('timezone') &&
         <Autocomplete
           options={timezones}
-          getOptionLabel={(timezone: timezoneType) => ( `${timezone.timezoneName} (GMT${timezone.timezoneOffset})` )}
-          defaultValue={selectedTimezone}
+          getOptionLabel={(timezone: timezoneType) => (`${timezone.timezoneName} (GMT${timezone.timezoneOffset})`)}
+          defaultValue={timezoneObj}
           onChange={handleTimezoneChange()}
```
Comment (rart) [inline ui/app/src/components/DateTimePicker.tsx]: No point for handler creator

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 10]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
 
 export const paths = (checked: any) => (
   Object.entries({ ...checked })
-    .filter(([key, value]) => value === true)
-    .map(([key]) => key)
+  .filter(([key, value]) => value === true)
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Did you import the code style settings? Missing indentation.

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 11]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
-    publishing: false,
     error: false,
     global: false,
     errorResponse: null
   });
 
-  const { formatMessage } = useIntl();
-
   const user = useSelector<GlobalState, GlobalState['user']>(state => state.user);
   const siteId = useSelector<GlobalState, GlobalState['sites']>(state => state.sites).active;
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: useActiveSite

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 12]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
+  const userSitesRoles: String[] = user.rolesBySite[siteId];    // TODO: check why is not working with editorial-ice
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Did you check? If this is not done then this PR is not complete. Otherwise, remove TODO.

What is it that’s “not working”?

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 13]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
+  const submit = submitMap[userRole];
+
+  const { formatMessage } = useIntl();
 
   useEffect(getPublishingChannels, []);
-  useEffect(setRef, [checkedItems, checkedSoftDep]);
+  useEffect(() => {
+    const result = (
+      Object.entries({ ...checkedItems, ...checkedSoftDep })
+      .filter(([key, value]) => value === true)
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Are there other truthy values coming here to use `... === true` instead of simply `(...) => value`

Missing indentation. Check code style settings.

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 14]
```diff
@@ -82,65 +89,94 @@ export const selectAllDeps = (setChecked: Function, items: Item[]) => {
...
-      .subscribe(
-        ({ response }) => {
-          setPublishingChannels(response.availablePublishChannels);
-          setPublishingChannelsStatus('Success');
-          setSubmitDisabled(false);
-        },
-        ({ response }) => {
-          setPublishingChannelsStatus('Error');
-          setSubmitDisabled(true);
+    .subscribe(
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: move .subscribe( to same line as the fetch invocation

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 15]
```diff
@@ -202,56 +237,51 @@ function RequestPublishDialog(props: RequestPublishDialogProps) {
...
-          });
-        },
-        () => {
-          setDeps({
-            items1: [],
-            items2: []
-          });
-        }
-      );
+    .subscribe(
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: fetch.. subscribe( same line

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 16]
```diff
@@ -202,56 +237,51 @@ function RequestPublishDialog(props: RequestPublishDialogProps) {
...
       handleClose={handleClose}
       handleSubmit={handleSubmit}
       submitDisabled={submitDisabled}
+      setSubmitDisabled={setSubmitDisabled}
+      showDepsDisabled={showDepsDisabled}
       dialog={dialog}
       setDialog={setDialog}
       open={open}
-      title={formatMessage(messages.title)}
+      title={userRole === 'admin' ? formatMessage(goLiveMessages.title) : formatMessage(submitMessages.title)}
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: move ternary inside the formatMessage

File: ui/app/src/modules/Content/Publish/PublishForm.tsx (hunk 6 above)
Comment (rart) [reply ui/app/src/modules/Content/Publish/PublishForm.tsx]: Timeout should be cancelled if the radio goes back to now

**Review by rart — APPROVED** [review]

**Review by rart — CHANGES_REQUESTED** [review]
(Pressed wrong radio on review submission)

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx (hunk 10 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishDialog.tsx]: Yes it is updated, that's weird

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx (hunk 13 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishDialog.tsx]: No values different than boolean, removing

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx (hunk 12 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishDialog.tsx]: removing comment, it is not working on node server but it is in studio after build

File: ui/app/src/modules/Content/Publish/PublishDialog.tsx [hunk 17]
```diff
@@ -85,62 +93,87 @@ export const paths = (checked: any) => (
...
     error: false,
     global: false,
     errorResponse: null
   });
 
-  const { formatMessage } = useIntl();
-
   const user = useSelector<GlobalState, GlobalState['user']>(state => state.user);
-  const siteId = useSelector<GlobalState, GlobalState['sites']>(state => state.sites).active;
+  // const siteId = useSelector<GlobalState, GlobalState['sites']>(state => state.sites).active;
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishDialog.tsx]: x

File: ui/app/src/modules/Content/Publish/PublishForm.tsx [hunk 18]
```diff
@@ -103,7 +103,7 @@ const useStyles = makeStyles(() => createStyles({
   },
   selectInput: {
     padding: '10px 12px',
-    backgroundColor: '#fff'
+    backgroundColor: palette.white
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishForm.tsx:106]: Is this not the default bg?

File: ui/app/src/modules/Content/Publish/PublishForm.tsx [hunk 19]
```diff
@@ -145,7 +145,7 @@ const useStyles = makeStyles(() => createStyles({
     paddingBottom: 0
   },
   textField: {
-    backgroundColor: '#ffffff',
+    backgroundColor: palette.white,
```
Comment (rart) [inline ui/app/src/modules/Content/Publish/PublishForm.tsx:148]: Is this not the default bg?

File: ui/app/src/modules/Content/Publish/PublishForm.tsx (hunk 18 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishForm.tsx:106]: Nope, the default is a grey

File: ui/app/src/modules/Content/Publish/PublishForm.tsx (hunk 19 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Publish/PublishForm.tsx:148]: Nope, the default is a grey

**Review by rart — APPROVED** [review]
